src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp

```diff
@@ -81,25 +82,33 @@ ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::ShenandoahClassLoad
   if (!SINGLE_THREADED) {
     ClassLoaderDataGraph::clear_claimed_marks();
   }
-  if (CONCURRENT) {
+  if (CONCURRENT && !SINGLE_THREADED) {
     ClassLoaderDataGraph_lock->lock();
   }
 }
 
 template <bool CONCURRENT, bool SINGLE_THREADED>
 ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::~ShenandoahClassLoaderDataRoots() {
-  if (CONCURRENT) {
+  if (CONCURRENT && !SINGLE_THREADED) {
     ClassLoaderDataGraph_lock->unlock();
   }
 }
 
-
 template <bool CONCURRENT, bool SINGLE_THREADED>
 void ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::always_strong_cld_do(CLDClosure* clds, uint worker_id) {
   if (SINGLE_THREADED) {
-    assert(SafepointSynchronize::is_at_safepoint(), "Must be at a safepoint");
-    assert(Thread::current()->is_VM_thread(), "Single threaded CLDG iteration can only be done by VM thread");
-    ClassLoaderDataGraph::always_strong_cld_do(clds);
+    if (CONCURRENT) {
+      if (_semaphore.try_acquire()) {
+        ShenandoahWorkerTimingsTracker timer(_phase, ShenandoahPhaseTimings::CLDGRoots, worker_id);
+        MutexLocker locker(ClassLoaderDataGraph_lock, Mutex::_no_safepoint_check_flag);
+        ClassLoaderDataGraph::always_strong_cld_do(clds);
+        _semaphore.claim_all();
+      }
+    } else {
+      assert(SafepointSynchronize::is_at_safepoint(), "Must be at a safepoint");
+      assert(Thread::current()->is_VM_thread(), "Can only be done by VM thread");
+      ClassLoaderDataGraph::always_strong_cld_do(clds);
+    }
   } else if (_semaphore.try_acquire()) {
```

Comment:
It looks to me that we do:
if (_semaphore.try_acquire()) {

in both branches now. Can this be reshaped to do try_acquire() and claim_all() outermost?